WhatAndroid/src/main/java/what/whatandroid/views/TouchImageView.java

```diff
@@ -0,0 +1,335 @@
+package what.whatandroid.views;
+
+import android.content.Context;
+import android.graphics.Matrix;
+import android.graphics.PointF;
+import android.graphics.drawable.Drawable;
+import android.util.AttributeSet;
+import android.util.Log;
+import android.view.MotionEvent;
+import android.view.ScaleGestureDetector;
+import android.view.View;
+import android.widget.ImageView;
+
+public class TouchImageView extends ImageView {
+	Matrix matrix;
+	// We can be in one of these 3 states
+	static final int NONE = 0;
+	static final int DRAG = 1;
+	static final int ZOOM = 2;
+
+	int mode = NONE;
+
+	// Remember some things for zooming
+	PointF last = new PointF();
+	PointF start = new PointF();
+	float minScale = 1f;
+	float maxScale = 3f;
+	float[] m;
+	int viewWidth, viewHeight;
+
+	static final int CLICK = 3;
+
+	float saveScale = 1f;
+
+	protected float origWidth, origHeight;
+
+	int oldMeasuredWidth, oldMeasuredHeight;
+
+	ScaleGestureDetector mScaleDetector;
+
+	Context context;
+
+	public TouchImageView(Context context) {
+		super(context);
+		sharedConstructing(context);
+	}
+
+	public TouchImageView(Context context, AttributeSet attrs) {
+		super(context, attrs);
+		sharedConstructing(context);
+	}
+
+	private void sharedConstructing(Context context) {
```

Comment:
For readability I'd prefer it if there wasn't a newline after every statement here but that things were grouped logically instead.